### PR TITLE
dtls.c: Check length of data for ServerHello

### LINKED ISSUE #55: Buffer Over-read Affecting check_server_hello()
Hello again,

The following over-read happens when you send a fragmented Server Hello message to the client. The bug is found employing symbolic execution technique. Suppose you have a `Server Hello` message with the following values for certain fields:

- `length` in the record layer = 46
- `length` in the handshake layer = 0
- `fragment length` = 128

When we are handling the message (in `dtls_handle_message()`), `data_length = decrypt_verify(peer, msg, rlen, &data)` is called. It returns a pointer to the first offset of the handshake layer (`uint8 *data`) and the real size of the handshake layer (`data_length` = 46). Later inside `check_server_hello(ctx, peer, data, data_length)`, Tiny tries to read different fields of the `Server Hello` message. It happens by advancing the pointer `data` and deducting the `data_length`. As I previously mentioned, the `Server Hello` is fragmented and it contains only up to the `random` field (`handshake header = 12` bytes + `handshake version = 2` bytes + `random = 32` bytes). Therefore, after advancing the pointer (`data`) for 46 bytes, Tiny tries to skip the `Session ID` by calling the [Macro](https://github.com/eclipse/tinydtls/blob/feature/handshake_fragmentation/dtls.c#L2765) (`SKIP_VAR_FIELD(data, data_length, uint8)`). However, Tiny already has advanced the pointer (`data`) for 46 bytes, and the pointer (`data`) is currently pointing to invalid memory. Therefore, here we have a one-byte over-read. I have added the assertion (`assert(data_length > 0)`) before `SKIP_VAR_FIELD(data, data_length, uint8)` to reveal the invalid read. For reproducing the bug, I have written a small UDP server that plays the role of a DTLS-server. 
I have attached the means to reproduce the mentioned bug. To do so, after downloading the suite, in tinydtls-witness/tests, execute the script `setupserver.sh` to compile TinyDTLS and `simple-udp-server`. Note that, I have de-randomized TinyDTLS for this demonstration and the patch is available in `tinydtls-witness/tests/patches`. Successful execution of `setupserver.sh` will run `simple-udp-server` on port 20220. Now, if you execute `reproduce.sh`, it will initiate a handshake with `simple-udp-server` and when it gets the Server Hello, the assertion that I added will fail.

image

I hope I could explain it understandably. I appreciate the confirmation of the reported bugs.

Best,
Hooman Asadian
[tinydtls-witness.zip](https://github.com/eclipse/tinydtls/files/6182299/tinydtls-witness.zip)

### PR DESCRIPTION
**Pull request by mrdeep1**
If insufficient data, alert out.

Fixes #55.

Signed-off-by: Jon Shallow <supjps-libcoap@jpshallow.com>

### CODE REVIEW

**Review by bathooman — APPROVED** [review]
The pull request successfully resolves the issue.

**Comment by boaks** [comment]
That PR runs also successful in Californium's interoperability tests.

File: dtls.c [hunk 1]
```diff
@@ -2766,18 +2766,14 @@ check_server_hello(dtls_context_t *ctx,
...
   dtls_handshake_parameters_t *handshake = peer->handshake_params;
 
-  /* This function is called when we expect a ServerHello (i.e. we
-   * have sent a ClientHello).  We might instead receive a HelloVerify
-   * request containing a cookie. If so, we must repeat the
-   * ClientHello with the given Cookie.
-   */
-  if (data_length < DTLS_HS_LENGTH + DTLS_HS_LENGTH)
+  /* Check we have enough data for the ServerHello */
+  if (data_length < DTLS_HS_LENGTH + 2 + DTLS_RANDOM_LENGTH + 1 + 2 + 1) {
```
Comment (obgm) [inline dtls.c:2770]: This test solves only part of the issue as the session id might not be empty. To handle that case, the data access after the `SKIP_VAR_FIELD()` for the session id must be protected as well.

File: dtls.c [hunk 2]
```diff
@@ -2766,18 +2766,14 @@ check_server_hello(dtls_context_t *ctx,
 {
   dtls_handshake_parameters_t *handshake = peer->handshake_params;
 
-  /* This function is called when we expect a ServerHello (i.e. we
-   * have sent a ClientHello).  We might instead receive a HelloVerify
-   * request containing a cookie. If so, we must repeat the
-   * ClientHello with the given Cookie.
-   */
-  if (data_length < DTLS_HS_LENGTH + DTLS_HS_LENGTH)
+  /* Check we have enough data for the ServerHello */
```
[the remark this answers is not in the document]
Comment (mrdeep1) [reply dtls.c:2769]: Comment updated

File: dtls.c (hunk 1 above)
Comment (mrdeep1) [reply dtls.c:2773]: Good catch.  Length rechecked after `SKIP_VAR_FIELD()` and code pushed.

**Pull request merged**